lxd/firewall/drivers/drivers_xtables.go

```diff
@@ -936,48 +940,53 @@ func (d Xtables) InstanceSetupProxyNAT(projectName string, instanceName string,
 
 	comment := d.instanceDeviceIPTablesComment(projectName, instanceName, deviceName)
 
-	targetRangeIndex := 0
-	for i := range forward.ListenPorts {
-		// Use the target port that corresponds to the listen port (unless only 1 is specified, in which
-		// case use the same target port for all listen ports).
-		targetIndex := 0
-		if targetPortsLen > 1 {
-			targetIndex = i
-		}
-
-		listenPortStr := fmt.Sprintf("%d", forward.ListenPorts[i])
-		targetPortStr := fmt.Sprintf("%d", forward.TargetPorts[targetIndex])
-
-		// Format the destination host/port as appropriate.
-		targetDest := fmt.Sprintf("%s:%s", targetAddressStr, targetPortStr)
-		if ipVersion == 6 {
-			targetDest = fmt.Sprintf("[%s]:%s", targetAddressStr, targetPortStr)
-		}
+	for i, listenPortRange := range forward.ListenPortRanges {
+		targetPortRange := forward.TargetPortRanges[i]
```

Comment:
@markylaing I think we might have problems with expecting the target ranges groups to match to the listen range groups.

For example this is currently allowed:

```
lxc config device set c1 p1  nat=true listen=tcp:192.168.1.200:80,81-82 connect=tcp:0.0.0.0:80-81,82
```

Which is equivalent to:

```
lxc config device set c1 p1  nat=true listen=tcp:192.168.1.200:80-82 connect=tcp:0.0.0.0:80-82
lxc config device set c1 p1  nat=true listen=tcp:192.168.1.200:80-81,82 connect=tcp:0.0.0.0:80-81,82
```

So we need to have a think about how can parse the config to its most efficient "canonical" form before passing the ranges around internally to each package.

Reply:
I had thought of this also. My plan was to sort a list of ports and then determine the ranges from there. However, this would mean that e.g. `9000-9001,80-81` would become `80-81,9000-9001`. This becomes problematic if we expect the given port ranges to correspond.

Reply:
@tomponline I've been around the houses with this for a little while now.

I think that we should keep the behaviour of simply passing around a slice of ports. This is because ranges calculated from the slice are automatically normalised (see https://github.com/lxc/lxd/pull/9614/commits/823b9d663e05fe5cf6f6fd7dd9b39ef8906f8dec). We can calculate the ranges where we need them and know that they are in their "canonical" form. If we pass around port ranges, it might become a bit less obvious whether the ranges are normalised or not. This also reduces the number of changes that need to made to various data types across the codebase.

Reply:
@markylaing sounds like it makes sense. So the listen and target port slices are indexed in the correct order already aren't they? So then the firewall driver then needs to detect consecutive ranges in those lists and compress them into fewer rules where it can.

Reply:
Yes they are indexed in the correct order. There are tests in https://github.com/lxc/lxd/blob/master/lxd/main_forkproxy_test.go though actually all of the cases have ports in ascending order. I've also added tests for the conversion of the port slice into the port range slice.

The xtables driver on this branch is detecting the ranges and compressing the rules now. It falls back to the original behaviour if an equal number of ports are supplied but it can't optimise the rules.